server/src/main/java/org/elasticsearch/cluster/metadata/FieldInferenceMetadata.java

```diff
@@ -0,0 +1,227 @@
+
+            // Get or create the set associated with the inferenceId
+            Set<String> fields = fieldsForInferenceIdsMap.computeIfAbsent(inferenceId, k -> new HashSet<>());
+            fields.add(fieldName);
+        }
+
+        fieldsForInferenceIds = Collections.unmodifiableMap(fieldsForInferenceIdsMap);
+        return fieldsForInferenceIds;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        FieldInferenceMetadata that = (FieldInferenceMetadata) o;
+        return Objects.equals(fieldInferenceMap, that.fieldInferenceMap);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(fieldInferenceMap);
+    }
+
+    @Override
+    public Diff<FieldInferenceMetadata> diff(FieldInferenceMetadata previousState) {
+        if (previousState == null) {
+            previousState = EMPTY;
+        }
+        return new FieldInferenceMetadataDiff(previousState, this);
+    }
+
+    static class FieldInferenceMetadataDiff implements Diff<FieldInferenceMetadata> {
+
+        public static final FieldInferenceMetadataDiff EMPTY = new FieldInferenceMetadataDiff(
+            FieldInferenceMetadata.EMPTY,
+            FieldInferenceMetadata.EMPTY
+        );
+
+        private final Diff<ImmutableOpenMap<String, FieldInference>> fieldInferenceMapDiff;
+
+        private static final DiffableUtils.DiffableValueReader<String, FieldInference> FIELD_INFERENCE_DIFF_VALUE_READER =
+            new DiffableUtils.DiffableValueReader<>(FieldInference::new, FieldInferenceMetadataDiff::readDiffFrom);
+
+        FieldInferenceMetadataDiff(FieldInferenceMetadata before, FieldInferenceMetadata after) {
+            fieldInferenceMapDiff = DiffableUtils.diff(
+                before.fieldInferenceMap,
+                after.fieldInferenceMap,
+                DiffableUtils.getStringKeySerializer(),
+                FIELD_INFERENCE_DIFF_VALUE_READER
+            );
+        }
+
+        FieldInferenceMetadataDiff(StreamInput in) throws IOException {
+            fieldInferenceMapDiff = DiffableUtils.readImmutableOpenMapDiff(
+                in,
+                DiffableUtils.getStringKeySerializer(),
+                FIELD_INFERENCE_DIFF_VALUE_READER
+            );
+        }
+
+        public static Diff<FieldInference> readDiffFrom(StreamInput in) throws IOException {
+            return SimpleDiffable.readDiffFrom(FieldInference::new, in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            fieldInferenceMapDiff.writeTo(out);
+        }
+
+        @Override
+        public FieldInferenceMetadata apply(FieldInferenceMetadata part) {
+            return new FieldInferenceMetadata(fieldInferenceMapDiff.apply(part.fieldInferenceMap));
+        }
+    }
+
+    public record FieldInference(String inferenceId, Set<String> sourceFields)
```

Comment:
Maybe  rename into `FieldInferenceOptions` since we'll probably need to add more later (shouldFailOnError, chunking, ...)?

Reply:
Makes sense :+1